ansible/roles/cri/tasks/containerd.yml

```diff
@@ -1,98 +1,63 @@
-    state: present
-  when: ansible_os_family | lower == 'debian'
-
-- name: 'debian : add apt repository for stable version'
-  apt_repository:
-    repo: deb https://download.docker.com/linux/ubuntu {{ ansible_lsb.codename }} stable
-    state: present
-  when: ansible_os_family | lower == 'debian'
-
-- name: 'debian : install containerd package'
-  apt:
-    name:
-      - containerd.io
-    state: present
-    update_cache: yes
-  when: ansible_os_family | lower == 'debian'
-
-- name: 'archlinux : install containerd package'
+- name: 'APT Installation'
+  when: ansible_os_family|lower == 'debian'
+  block:
+  - name: 'containerd | ensure docker.io apt repository public key is installed'
+    ansible.builtin.apt_key:
+      url: '{{ cri_containerd_repo_gpgkey }}'
+  - name: 'containerd | ensure docker.io apt repository is enabled'
+    ansible.builtin.apt_repository:
+      repo: 'deb {{ cri_containerd_repo_base_url }} {{ ansible_distribution_release }} stable'
+      state: 'present'
+  # this is to prevent apt upgrade from breaking kubernetes
+  - name: 'containerd | pin package version to prevent os upgrade clubbing'
+    ansible.builtin.template:
+      src: 'apt-preferences.d/containerd.j2'
+      dest: '/etc/apt/preferences.d/containerd'
+      owner: 'root'
+      mode: 0644
+  - name: 'containerd | install package from apt repository'
+    ansible.builtin.apt:
+      name:
+        - '{{ cri_containerd_package_name[cri_containerd_version | string] }}'
+      state: 'present'
+      update_cache: true
+      cache_valid_time: 3600
+
+# Archlinux installation of containerd package
+- name: 'containerd | install containerd package'
   pacman:
-    name:
-      - containerd
+    name: containerd
     state: present
     update_cache: yes
   when: ansible_os_family | lower == 'archlinux'
 
-- name: create containerd config directory
-  become: true
-  file:
-    path: /etc/containerd/
-    state: directory
-    owner: root
-    group: root
-    mode: '0755'
-
-# TODO: https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd-systemd
-- name: configure containerd
-  command:
-    cmd: containerd config default > /etc/containerd/config.toml
-    creates: /etc/containerd/config.toml
-  notify: restart containerd
-
-- name: start containerd
-  systemd:
-    name: containerd
-    daemon_reload: yes
-    state: started
-    enabled: yes
+- name: 'containerd | ensure "{{ cri_containerd_config_dir }}" config directory'
+  ansible.builtin.file:
+    path: '{{ cri_containerd_config_dir }}'
+    state: 'directory'
+    mode: 0755
+    owner: 'root'
+    group: 'root'
+
+# This is an idempotent check to ensure that the default configuration is present
+- name: 'containerd | check existing config.toml file'
+  ansible.builtin.stat:
+    path: '{{ cri_containerd_config_dir }}/config.toml'
+  register: 'containerd_config_file'
+
+- name: 'containerd | generate configuration file'
+  ansible.builtin.shell: |
```

Comment:
I think `ansible.builtin.command` is preferred over `ansible.builtin.shell` I'd need to reread what the justification was though.

Reply:
In this usage you're correct, `command` is preferred - the original task had a the config piped to it's final location.